group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -874,4 +1045,1265 @@ public void replay(
+            log.error("Group {} could not complete rebalance because no members rejoined.", groupId);
+
+            return tryCompleteJoinElseSchedule(group);
+        } else {
+            group.initNextGeneration();
+            if (group.isInState(EMPTY)) {
+                log.info("Group {} with generation {} is now empty ({}-{})",
+                    groupId, group.generationId(), topicPartition.topic(), topicPartition.partition());
+
+                CompletableFuture<Errors> appendFuture = new CompletableFuture<>();
+                appendFuture.whenComplete((error, t) -> {
+                    if (t != null || error != Errors.NONE) {
+                        // We failed to write the empty group metadata. If the broker fails before another rebalance,
+                        // the previous generation written to the log will become active again (and most likely timeout).
+                        // This should be safe since there are no active members in an empty generation, so we just warn.
+                        String message = t != null ? t.getMessage() : error.message();
+                        log.warn("Failed to write empty metadata for group {}: {}", group.groupId(), message);
+                    }
+                });
+
+                List<Record> records = Collections.singletonList(
+                    RecordHelpers.newGroupMetadataRecord(group, metadataImage.features().metadataVersion())
+                );
+
+                return new CoordinatorResult<>(records, appendFuture);
+
+            } else {
+                log.info("Stabilized group {} generation {} ({}-{}) with {} members",
+                    groupId, group.generationId(), topicPartition.topic(),
+                    topicPartition.partition(), group.size());
+
+                // Complete the awaiting join group response future for all the members after rebalancing
+                group.allMembers().forEach(member -> {
+
+                    List<JoinGroupResponseData.JoinGroupResponseMember> members = Collections.emptyList();
+                    if (group.isLeader(member.memberId())) {
+                        members = group.currentGenericGroupMembers();
+                    }
+
+                    JoinGroupResponseData response = new JoinGroupResponseData()
+                        .setMembers(members)
+                        .setMemberId(member.memberId())
+                        .setGenerationId(group.generationId())
+                        .setProtocolName(group.protocolName().orElse(null))
+                        .setProtocolType(group.protocolType().orElse(null))
+                        .setLeader(group.leaderOrNull())
+                        .setSkipAssignment(false)
+                        .setErrorCode(Errors.NONE.code());
+
+                    group.completeJoinFuture(member, response);
+                    rescheduleGenericGroupMemberHeartbeat(group, member);
```

Comment:
so that i understand: during completing rebalance, we schedule both pending sync (rebalance timeout) and heartbeats (session timeout). in practice, session timeout << rebalance timeout so heartbeats would expire and remove members. With cooperative rebalancing members should still be able to fetch records during completing rebalance phase. We want to extend the heartbeat here to rebalance timeout so that members are not removed by session timeout.

is this correct?

If so, i agree removing the heartbeat schedule sounds like the best approach since members are removed when pending sync expires anyways.

Reply:
are you referring to any lingering heartbeats at this point? i think we can just cancel them here right